Woo Subscriptions Segment [MAILPOET-3471]

costasovo opened the pull request:
This PR adds a new dynamic segment for active WooCommerce subscriptions.

I tried using [wcs_get_subsciptions](https://github.com/wp-premium/woocommerce-subscriptions/blob/master/wcs-functions.php#L425) for the implementation of the subscriber filter but since it does additional 3 queries per subscriber I decided to go with direct SQL.

The PR also adds a basic infrastructure for working with 3rd party plugins.

[MAILPOET-3471]

[MAILPOET-3471]: https://mailpoet.atlassian.net/browse/MAILPOET-3471

costasovo commented on assets/js/src/segments/dynamic/form.tsx:
```diff
@@ -58,6 +59,12 @@ function getAvailableFilters(): GroupFilterValue[] {
       options: WooCommerceOptions,
     });
   }
+  if (MailPoet.isWoocommerceActive) {
```
I added the check in a [later commit](https://github.com/mailpoet/mailpoet/pull/3451/commits/9e52571657b83a2faa500e6e1bc74d508b58ded9).

costasovo commented on lib/Segments/DynamicSegments/Filters/WooCommerceSubscription.php:
```diff
@@ -0,0 +1,45 @@
+use MailPoetVendor\Doctrine\DBAL\Query\QueryBuilder;
+use MailPoetVendor\Doctrine\ORM\EntityManager;
+
+class WooCommerceSubscription implements Filter {
```
We discussed it on Slack. The thing is that integration tests infrastructure doesn't have support for WooCommerce and WooCommerce subscriptions and it is not easy to add because of [the serialization issues](https://mailpoet.atlassian.net/browse/MAILPOET-3274).

costasovo commented on lib/Segments/DynamicSegments/Filters/WooCommerceSubscription.php:
```diff
@@ -0,0 +1,45 @@
+    $filterData = $filter->getFilterData();
+    $productId = (int)$filterData->getParam('product_id');
+    $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+    return $queryBuilder->innerJoin(
```
I unified the code formatting. As for the alias, we have to use the subscribers' table since the $queryBuilder that is passed as an argument is expected to select from the subscribers' table.

costasovo commented:
@pavel-mailpoet I added 3 new commits addressing your comments.

veljkho reviewed (changes requested):
Everything's good except small thing:
- When I deactivate WooCommerce plugin only (leaving Subscription plugin as active), I could still see total amount of users using subscriptions in the Segments Lists page... and no message that I should have WooCommerce installed in order to see total number calculation... According to the specs.. The both plugins must be activated in order to see calculations/edit segment etc.

Screenshot of issue without WC activated: https://d.pr/i/OK21Fb
However, message is present if I also deactivate Subscriptions plugin: https://d.pr/i/I9KV5E

costasovo commented:
I added 2 new commits:

- update of form component so that it is up to date with recent changes done in https://github.com/mailpoet/mailpoet/pull/3456/files
- fixed the issue reported by QA
